Mathlib/FieldTheory/KummerExtension.lean

```diff
@@ -0,0 +1,444 @@
+
+lemma autEquivRootsOfUnity_apply_rootOfSplit (hn : 0 < n) (e : L ≃ₐ[K] L) :
+    e (rootOfSplitsXPowSubC n a L hn) =
+      autEquivRootsOfUnity hζ H L hn e • (rootOfSplitsXPowSubC n a L hn) := by
+  obtain ⟨η, rfl⟩ := (autEquivRootsOfUnity hζ H L hn).symm.surjective e
+  rw [MulEquiv.apply_symm_apply, autEquivRootsOfUnity]
+  simp only [MulEquiv.symm_trans_apply, AlgEquiv.autCongr_symm, AlgEquiv.symm_symm,
+    MulEquiv.symm_symm, autAdjoinRootXPowSubCEquiv_apply, AlgEquiv.autCongr_apply,
+    AlgEquiv.trans_apply, adjoinRootXPowSubCEquiv_symm_eq_root, adjoinRootXPowSubCEquiv_root,
+    autAdjoinRootXPowSubC_root, AlgEquiv.map_smul]
+  rfl
+
+lemma autEquivRootsOfUnity_smul (hn : 0 < n) (e : L ≃ₐ[K] L)
+    (α : L) (hα : α ^ n = algebraMap K L a) :
+    autEquivRootsOfUnity hζ H L hn e • α = e α := by
+  have ⟨ζ, hζ'⟩ := hζ
+  rw [mem_primitiveRoots hn] at hζ'
+  have := nthRoots_eq_of_isPrimitiveRoot hn (hζ'.map_of_injective (algebraMap K L).injective)
+    (rootOfSplitsXPowSubC_pow n a L hn)
+  rw [← mem_nthRoots hn] at hα
+  simp only [this, Finset.range_val, Multiset.mem_map, Multiset.mem_range] at hα
+  obtain ⟨i, _, rfl⟩ := hα
+  simp only [map_mul, ← map_pow, ← Algebra.smul_def, AlgEquiv.map_smul,
+    autEquivRootsOfUnity_apply_rootOfSplit hζ H L]
+  exact smul_comm _ _ _
+
+lemma ext_of_isSplittingField_X_pow_sub_C (e₁ e₂ : L ≃ₐ[K] L)
+    (α : L) (hα : α ^ n = algebraMap K L a) (h : e₁ α = e₂ α) : e₁ = e₂ := by
+  have hn := Nat.pos_of_ne_zero (ne_zero_of_irreducible_X_pow_sub_C H)
+  apply (autEquivRootsOfUnity hζ H L hn).injective
+  by_cases hn' : n = 1
+  · subst hn'
+    apply (config := {allowSynthFailures := true }) Subsingleton.elim
+    exact ⟨fun a b ↦ by simpa using a.prop.trans b.prop.symm⟩
+  simp_rw [← autEquivRootsOfUnity_smul hζ H L hn _ α hα, Subgroup.smul_def] at h
+  have : α ≠ 0
+  · rintro rfl
+    apply ne_zero_of_irreducible_X_pow_sub_C' hn' H
+    rwa [zero_pow hn, eq_comm, map_eq_zero_iff _ (algebraMap K L).injective] at hα
+  ext
+  exact smul_left_injective _ this h
+
+/-- Suppose `L/K` is the splitting field of `Xⁿ - a`, and `ζ` is a `n`-th primitive root of unity
+in `K`, then `Gal(L/K)` is isomorphic to `ZMod n`. -/
+noncomputable
+def autEquivZmod {ζ : K} (hζ : IsPrimitiveRoot ζ n) :
+    (L ≃ₐ[K] L) ≃* Multiplicative (ZMod n) :=
+  haveI hn := Nat.pos_iff_ne_zero.mpr (ne_zero_of_irreducible_X_pow_sub_C H)
+  (autEquivRootsOfUnity ⟨ζ, (mem_primitiveRoots hn).mpr hζ⟩ H L hn).trans
+    ((MulEquiv.subgroupCongr (IsPrimitiveRoot.zpowers_eq (k := ⟨n, hn⟩)
+      (hζ.isUnit_unit' hn)).symm).trans (AddEquiv.toMultiplicative'
+        (hζ.isUnit_unit' hn).zmodEquivZPowers.symm))
+
+lemma autEquivZmod_symm_apply {ζ : K} (hζ : IsPrimitiveRoot ζ n)
+    (α : L) (hα : α ^ n = algebraMap K L a) (m : ℤ) :
+    (autEquivZmod H L hζ).symm (Multiplicative.ofAdd (m : ZMod n)) α = ζ ^ m • α := by
+  haveI hn := Nat.pos_iff_ne_zero.mpr (ne_zero_of_irreducible_X_pow_sub_C H)
```

Comment:
do you need all these `letI`'s and `haveI`'s to be the I variants, seems unnecessary to me

Reply:
I removed the `letI`s and `haveI`s in lemmas. However `letI` has the benefit of not needing to name instances, so I kept some of them.